Add API handlers for CRUD operations on conversation members

igorveselinovic opened the pull request:
Resolves #9 

Added five handlers for CRUD operations on the members in a conversation, including some database logic. Refactored some existing code here and there, too.

igorveselinovic committed:
Add empty CRUD handlers for user mappings

igorveselinovic committed:
Add new database functions for mappings

Functions for updating a single row and deleting a single row.

igorveselinovic committed:
Move mapping handlers to their own file

Rename handlers.go to differentiate between the two.

igorveselinovic committed:
Add POST handler for user mappings

Creates a single user-to-conversation mapping in the database.

igorveselinovic committed:
Add check for invalid role when creating mapping

Can only be "admin" or "user" since only the original conversation
creator can be an "owner".

igorveselinovic committed:
Add GET handler for user mappings

Moved the user mapping helper function to the mappings handler file,
since that makes more sense. Some of the code in the function had to be
taken out because the error response when a user mapping doesn't exist
isn't always the same. This unfortunately makes the code less DRY
overall.

igorveselinovic committed:
Refactor some code

* Use gorilla middleware model for logging
* Add helper function for internal server error response just to remove
  some repetition

igorveselinovic committed:
Make different handlers more consistent

igorveselinovic committed:
Add GET handler for getting multiple user mappings

Currently returns an unsorted list of all the user mappings that belong
to a single conversation. MySQL seems to sort them by UserID by default.
Would need to somehow communicate with Karen to get user names to allow
for alphabetical sorting of this list by ether.

igorveselinovic committed:
Add PATCH handler for conversation members

igorveselinovic committed:
Make variable names consistent

Use "member" as term for a user-conversation mapping

igorveselinovic committed:
Update member handlers to better fit requirements

Role setting/modification now follows requirements.

igorveselinovic committed:
Add check for MySQL duplicate entry error

This is checked in the POST handler for members, and is specially
handled to return a 409 conflict to indicate to the client that a user
has already been added to the conversation.

igorveselinovic committed:
Update DB queries to treat pending flag as int

They way it worked previously always resulted in GET requests saying
every member had a pending state of false.

igorveselinovic committed:
Add DELETE handler for conversation members

Added some Role utility functions to simplify working with them.

igorveselinovic committed:
Add deferred close of req body in PATCH handler

igorveselinovic committed:
Add DB schema file for the two tables

Thao-Tran commented on app/server.go:
```diff
@@ -35,22 +35,47 @@ func main() {
+		"/ether/v1/conversations/{conversation_id:[0-9]+}/users/{user_id:[0-9]+}",
+		env.DeleteMappingHandler,
+	).Methods("DELETE")
+	httpMux.Use(logging)
```
Oh damn this is a great idea!

Thao-Tran commented on handlers/conversation_handlers.go:
```diff
@@ -13,12 +13,7 @@ import (
-}
-
-func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Conversation) error {
+func parseConversationJSON(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Conversation) error {
```
You can combine this helper function and the `parseMappingJSON` by changing the type of `bodyObj` to `interface{}`.

Thao-Tran commented on handlers/conversation_handlers.go:
```diff
@@ -40,9 +35,7 @@ func parseReqBody(w http.ResponseWriter, body io.ReadCloser, bodyObj *models.Con
-		errMsg := "Internal Server Error"
-		log.Println(errMsg + ": " + err.Error())
-		http.Error(w, errMsg, http.StatusInternalServerError)
+		internalServerError(w, err)
```
Glad to see you figured out how to clear out some of the error handling repetition.

Thao-Tran commented on handlers/conversation_handlers.go:
```diff
@@ -140,8 +112,15 @@ func (env *Env) GetConversationHandler(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+	if mapping == nil {
+		errMsg := fmt.Sprintf("User %d is not in conversation %d", userID, conversationID)
```
Maybe you can make a helper function for this error handling since the same thing seems to be repeated a couple times (lines 164 and 229).

Thao-Tran commented on handlers/mapping_handlers.go:
```diff
@@ -0,0 +1,404 @@
+	}
+	memberList := &models.UserConversationMappingList{Users: members}
+
+	// TODO: Get user names from karen?
```
I think we can leave that for the client to do when its rendering the page.

igorveselinovic replied:
Yeah, forgot to elaborate on this. The question is if we want to do sorting on the front-end or back-end, because if we want the back-end to sort the list of users by something useful, we'd need to get that from `karen`.

Thao-Tran replied:
Oh true. Hmmm. I guess we'd probably be doing the same thing in either case, i.e. client makes the request to `ether`, `ether` gets the sorted users from `karen`, then `ether` returns the user-conversation mapping merged with the sorted users or client makes request to `ether` gets the user-conversation mapping, makes request to `karen` get the sorted users and then matches the mapping to the sorted users itself. I think it'd be better to do it in the front-end, so that we don't have to send a request to `ether` and `karen` everytime the sorting option changes.

Thao-Tran commented on handlers/mapping_handlers.go:
```diff
@@ -0,0 +1,404 @@
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+	if userID == memberID {
```
But what if the user just wants to modify their own nickname?

igorveselinovic replied:
Yeah, forgot about nicknames.

Thao-Tran commented on handlers/mapping_handlers.go:
```diff
@@ -0,0 +1,404 @@
+		return
+	}
+
+	if reqMember.Role == "" && reqMember.Nickname == nil {
```
The `Pending` and `LastOpened` fields should be updated using this endpoint too right?

igorveselinovic replied:
I thought that these 2 fields would get updated based on receiving messages on a queue, rather than from a client request. But I guess maybe it does make more sense for clients to update these things themselves.

Thao-Tran replied:
Hmm. That's a good point. I think `LastOpened` could be handled by a message queue so when a client closes a WebSocket, i.e. closes a conversation, `patches` could put a message on the queue to be like "this user closed the conversation" so `ether` would update it. `Pending` I think can be purely client, since if they have some pending invitation, when they accept, the client can just send a PATCH to update it.

Thao-Tran commented on handlers/mapping_handlers.go:
```diff
@@ -0,0 +1,404 @@
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+	if userID == memberID {
```
Can people not leave conversations?

igorveselinovic replied:
Ah yeah, this should only be if the user has the `owner` role.

Thao-Tran commented on models/user_conversation_mapping.go:
```diff
@@ -36,13 +42,63 @@ const (
 
+// Merge creates a new UserConversationMapping by copying the original mapping
+// and replacing its fields with the non-zero-value fields of a patch mapping
+func (m *UserConversationMapping) Merge(patch *UserConversationMapping) *UserConversationMapping {
```
Shouldn't `Pending` and `LastOpened` be updated too?

igorveselinovic replied:
Yeah they should be.

Thao-Tran commented on handlers/mapping_handlers.go:
```diff
@@ -0,0 +1,404 @@
+		return
+	}
+
+	if reqMember.Role != models.Admin && reqMember.Role != models.User ||
```
Might be better to use the `Valid` method for the `reqMember.Role` here and then check if the members roles follow our rules after to make it more clear/readable.

Thao-Tran commented on handlers/mapping_handlers.go:
```diff
@@ -0,0 +1,404 @@
+	}
+
+	if reqMember.Role != "" {
+		if reqMember.Role != models.Admin && reqMember.Role != models.User {
```
Might help with readability if you use the `Valid` method for the `Role` type and then the `reqMember.Role != models.Owner`.

Thao-Tran reviewed (changes requested):
Looks good. Just a couple comments.

igorveselinovic committed:
Do some recommended refactoring

* Make a single helper for parsing JSON
* Remake a helper for getting user-to-conversation mappings to save on
  repetitive code

igorveselinovic committed:
Update logic to satisfy requirements

* Allow users to change their own nickname
* Allow users to leave a conversation
* Allow users to change their own invitation status

igorveselinovic committed:
Add a check on session user for pending invitation

Affects a few different handlers. Users who haven't accepted their
invitation yet should not be able to do modify the conversation or its
members in any way.
